deepchem/feat/rdkit_grid_featurizer.py

```diff
@@ -0,0 +1,1266 @@
+  3.141592653589793
+  """
+  vector_i_u = _unit_vector(vector_i)
+  vector_j_u = _unit_vector(vector_j)
+  angle = np.arccos(np.dot(vector_i_u, vector_j_u))
+  if np.isnan(angle):
+    if (vector_i_u == vector_j_u).all():
+      return 0.0
+    else:
+      return np.pi
+  return angle
+
+
+def _hash_sybyl(sybyl, sybyl_types):
+  return (sybyl_types.index(sybyl))
+
+
+def _hash_ecfp(ecfp, power):
+  """
+  Returns an int of size 2^power representing that
+  ECFP fragment. Input must be a string.
+  """
+
+  md5 = hashlib.md5()
+  md5.update(ecfp)
+  digest = md5.hexdigest()
+  ecfp_hash = int(digest, 16) % (2**power)
+  return (ecfp_hash)
+
+
+def _hash_ecfp_pair(ecfp_pair, power):
+  """
+  Returns an int of size 2^power representing that
+  ECFP pair. Input must be a tuple of strings.
+  """
+  ecfp = "%s,%s" % (ecfp_pair[0], ecfp_pair[1])
+  md5 = hashlib.md5()
+  md5.update(ecfp)
+  digest = md5.hexdigest()
+  ecfp_hash = int(digest, 16) % (2**power)
+  return (ecfp_hash)
+
+
+def _compute_all_ecfp(mol, indices=None, degree=2):
+  """
+  For each atom:
+    Obtain molecular fragment for all atoms emanating outward to given degree.
+    For each fragment, compute SMILES string (for now) and hash to an int.
+    Return a dictionary mapping atom index to hashed SMILES.
+  """
+
+  ecfp_dict = {}
+  for i in range(mol.GetNumAtoms()):
+    if indices is not None and i not in indices:
+      continue
+    env = Chem.FindAtomEnvironmentOfRadiusN(mol, degree, i, useHs=True)
+    submol = Chem.PathToSubmol(mol, env)
+    smile = Chem.MolToSmiles(submol)
+    ecfp_dict[i] = "%s,%s" % (mol.GetAtoms()[i].GetAtomicNum(), smile)
+
+  return ecfp_dict
+
+
+def _compute_ecfp_features(system_ob, ecfp_degree, ecfp_power):
+  """Computes ECFP features for provided openbabel molecule.
+
+  Parameters:
+  -----------
+    system_ob: openbabel molecules
+      Molecule to featurize.
+    ecfp_degree: int
+      ECFP radius
+    ecfp_power: int
+      Number of bits to store ECFP features (2^ecfp_power will be length of
+      ECFP array)
+  Returns:
+  --------
+    ecfp_array: np.ndarray
+      Returns an array of size 2^ecfp_power where array at index i has a 1 if
+      that ECFP fragment is found in the molecule and array at index j has a 0
+      if ECFP fragment not in molecule.
+  """
+  bv = AllChem.GetMorganFingerprint(system_ob, 2)
+  return [int(bv.GetBit(x)) for x in range(bv.GetNumBits())]
+
+
+def _featurize_binding_pocket_ecfp(protein_xyz,
+                                   protein,
+                                   ligand_xyz,
+                                   ligand,
+                                   pairwise_distances=None,
+                                   cutoff=4.5,
+                                   ecfp_degree=2):
+  """Computes ECFP dicts for ligand and binding pocket of the protein.
+
+  Parameters
+  ----------
+  protein_xyz: np.ndarray
+    Of shape (N_protein_atoms, 3)
+  protein: PDB object (TODO(rbharath): Correct?)
```

Comment:
Yes, but this is the PDB object from `nnscore_utils.py` (IIRC). Would be cool to see if we can get this to an rdkit/mdtraj object instead.